### PR TITLE
Introduce `bindAll`

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related

- Related Issue  #
- Closes https://github.com/Effect-TS/effect/issues/3484

### CODE REVIEW

**Comment by gcanti** [comment]
is `bindAll` truly necessary when nesting can achieve pretty much the same result without introducing new APIs?

```ts
Effect.Do.pipe(
  Effect.bind("x", () => Effect.succeed(2)),
  Effect.bind("y", ({ x }) =>
    Effect.all({
      a: Effect.succeed(x),
      b: Effect.succeed("ops")
    }, { mode: "either" }))
)
```

vs

```ts
Effect.Do.pipe(
  Effect.bind("x", () => Effect.succeed(2)),
  Effect.bindAll(({ x }) => ({
    a: Effect.succeed(x),
    b: Effect.succeed("ops")
  }), { mode: "either" })
)
```

**Comment by tim-smart** [comment]
> is `bindAll` truly necessary when nesting can achieve pretty much the same result without introducing new APIs?

With `Effect.all` you will need to follow that up with merging `y` into the result. So the equivalent would be:

```ts
Effect.Do.pipe(
  Effect.bind("x", () => Effect.succeed(2)),
  Effect.flatMap((value) =>
    Effect.all(
      {
        a: Effect.succeed(value.x),
        b: Effect.succeed("ops"),
      },
      { concurrency: "unbounded" },
    ).pipe(Effect.map((result) => ({ ...value, ...result }))),
  ),
)
```

**Comment by KhraksMamtsov** [comment]
On the one hand, I agree with Guilio. On the other hand, I saw that Michael approved of the idea. I had time to contribute to the effect and practice with the types)

File: packages/effect/src/Effect.ts [hunk 1]
```diff
@@ -594,18 +594,23 @@ export declare namespace All {
...
+    readonly mode?: "default" | "validate" | "either" | undefined
+    readonly concurrentFinalizers?: boolean | undefined
+  }
+
   /**
    * @since 2.0.0
    */
   export type Return<
     Arg extends Iterable<EffectAny> | Record<string, EffectAny>,
-    O extends {
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:602]: I think this should be reverted, and then remove the `discard` option.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/effect/src/Effect.ts (hunk 1 above)
Comment (mikearnaldi) [reply packages/effect/src/Effect.ts:600]: we also need the "concurrentFinalizers" option

**Comment by KhraksMamtsov** [comment]
@tim-smart, please confirm that everything is fine

File: packages/effect/src/Effect.ts [hunk 2]
```diff
@@ -3540,6 +3540,91 @@ export const bind: {
...
+ *   Effect.Do,
+ *   Effect.bind("x", () => Effect.succeed(2)),
+ *   Effect.bindAll(({ x }) => ({
+ *     a: Effect.succeed(x),
+ *     b: Effect.fail('ops'),
+ *   }), { concurrency: 2, mode: 'either' })
+ * )
+ * assert.deepStrictEqual(Effect.runSync(result), { x: 2, a: Either.right(2), b: Either.left('ops') })
+ *
+ * @see All.Options
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:3561]: Can remove this

File: packages/effect/src/Effect.ts [hunk 3]
```diff
@@ -3540,6 +3540,91 @@ export const bind: {
...
+  >(
+    f: (a: A) => [Extract<keyof X, keyof A>] extends [never] ? X : `Duplicate keys`,
+    options?: undefined | O
+  ): <E1, R1>(
+    self: Effect<A, E1, R1>
+  ) => Effect<
+    {
+      [K in keyof X | keyof A]: K extends keyof A ? A[K] :
+        K extends keyof Effect.Success<
+          All.ReturnObject<X, All.IsDiscard<O>, All.ExtractMode<O>>
```
Comment (tim-smart) [inline packages/effect/src/Effect.ts:3584]: You can remove the discard checks here.

**Review by tim-smart — CHANGES_REQUESTED** [review]

**Review by tim-smart — APPROVED** [review]

**Pull request merged**